src/app/utils/undo-redo-history.ts

```diff
@@ -0,0 +1,103 @@
+import { useCallback, useReducer } from 'react';
+
+interface History<S> {
+  state: S;
+  undoStack: S[];
+  redoStack: S[];
+}
+
+type StateUpdateFunction<S> = (oldState: S) => S;
+
+interface SetAction<S> {
+  type: 'set';
+  update: StateUpdateFunction<S>;
+}
+
+interface UndoAction {
+  type: 'undo';
+}
+
+interface RedoAction {
+  type: 'redo';
+}
+
+type Action<S> = SetAction<S> | UndoAction | RedoAction;
+
+function historyReducer<S>(oldState: History<S>, action: Action<S>): History<S> {
+  switch (action.type) {
+    case 'set': {
+      const { undoStack, state } = oldState;
+      return {
+        state: action.update(state),
+        undoStack: [...undoStack, state],
+        redoStack: [],
+      };
+    }
+    case 'undo': {
+      const { undoStack, redoStack, state } = oldState;
+      if (undoStack.length < 1) {
+        throw new Error("Can't undo");
+      }
+      const previousState = undoStack[undoStack.length - 1];
+      return {
+        state: previousState,
+        undoStack: undoStack.slice(0, -1),
+        redoStack: [...redoStack, state],
+      };
+    }
+    case 'redo': {
+      const { undoStack, redoStack, state } = oldState;
+      if (redoStack.length < 1) {
+        throw new Error("Can't redo");
+      }
+      const nextState = redoStack[redoStack.length - 1];
+      return {
+        state: nextState,
+        undoStack: [...undoStack, state],
+        redoStack: redoStack.slice(0, -1),
+      };
+    }
+  }
+}
+
+function initializer<S>(state: S): History<S> {
+  return {
+    state,
+    undoStack: [],
+    redoStack: [],
+  };
+}
+
+export function useHistory<S>(initialState: S): {
+  state: S;
+  setState: (f: StateUpdateFunction<S>) => void;
+  undo: () => void;
+  redo: () => void;
+  canUndo: boolean;
+  canRedo: boolean;
+} {
+  // Needed for type checking, TS otherwise seems to get lost
+  // in weaker overloads of `useReducer`?
+  const reducer: typeof historyReducer<S> = historyReducer;
+  const [{ state, undoStack, redoStack }, dispatch] = useReducer(
+    reducer,
+    initialState,
+    initializer
+  );
```

Comment:
I did not expect this to be valid syntax, and neither does VS Code's syntax highlighting
![grafik](https://user-images.githubusercontent.com/14299449/198605209-1b823abf-fbda-4c3a-85bc-5018512fef55.png)
there's gotta be a better way to do it but I couldn't find it

Reply:
Hm yeah that's unfortunate. I'm not sure why it has trouble with that typing.

Reply:
So I did some research and while we probably can't fix type inference/useReducer overloads, the solution I've used here is an application of TS 4.7 instantiation expressions. What I had tried first was `useReducer(historyReducer<S>, ...)`, which ought to work according to that feature's documentation, but there are bugs in VS Code that make constructs like these cause spurious errors:

https://github.com/microsoft/TypeScript/issues/50191
https://github.com/microsoft/TypeScript/issues/50161

It looks like the `typeof` workaround throws off the parser enough for it to not issue these kinds of semantic errors 🤷